grails-app/services/com/unifina/service/CanvasService.groovy

```diff
@@ -220,19 +222,17 @@ class CanvasService {
 					c.exampleType = ExampleType.NOT_SET
 					c.save(validate: true, failOnError: true)
 					permissionService.systemGrant(user, c, Permission.Operation.SHARE)
-
-					Map signalPathMap = (Map) JSON.parse(example.json)
-					resetUiChannels(signalPathMap)
-					SignalPathService.ReconstructedResult result = reconstructFrom(signalPathMap, user)
-					result.signalPath.setCanvas(c)
-					result.signalPath.getModules().each {
-						if (it instanceof ModuleWithUI) {
-							it.ensureUiChannel()
-						}
-					}
-					result.signalPath.ensureUiChannel()
-					c.json = example.json
-					c.save(validate: true, failOnError: true)
+					permissionService.systemGrant(user, c, Permission.Operation.READ)
+					permissionService.systemGrant(user, c, Permission.Operation.WRITE)
+
+					Map map = (JSONObject) JSON.parse(example.json)
+					SaveCanvasCommand cmd = new SaveCanvasCommand(
+						name: c.name,
+						modules: map?.getJSONArray("modules") == null ? [] : map?.getJSONArray("modules"),
+						settings: map?.getJSONObject("settings") == null ? [:] : map?.getJSONObject("settings"),
+						adhoc: false,
+					)
+					updateExisting(c, cmd, user,true)
```

Comment:
`updateExisting` saves the canvas AND grants the user all permissions to it, if it is a previously unsaved canvas. So for `ExampleType.COPY` it should be enough to:

- setup the `Canvas` object
- setup the `SaveCanvasCommand` object
- `updateExisting`
- nothing else.

So I think you don't need the `c.save(...)` and `permissionService.systemGrant(...)`s.